tools/ci/actions/check-configure.sh

```diff
@@ -0,0 +1,95 @@
+#*                                                                        *
+#*                 David Allsopp, OCaml Labs, Cambridge.                  *
+#*                                                                        *
+#*   Copyright 2021 David Allsopp Ltd.                                    *
+#*                                                                        *
+#*   All rights reserved.  This file is distributed under the terms of    *
+#*   the GNU Lesser General Public License version 2.1, with the          *
+#*   special exception on linking described in the file LICENSE.          *
+#*                                                                        *
+#**************************************************************************
+
+# Hygiene Checks: ensure that configure.ac generates configure
+# This tests both branches and PRs. Any commit which updates either files which
+# affect configure (configure.ac, VERSION, aclocal.m4 and build-aux/*) and also
+# which alter this script.
+# The behaviour is slightly different for pushes vs pull requests: in a PR, all
+# commits must be correct; in a push, it must be the case that the configure is
+# correct at the tip of the branch. This allows you to push a correcting PR to
+# trunk, for example, but rejects a PR which includes bad commits (for increased
+# bisect safety).
+
+set -e
+
+if [[ $1 = 'pull_request' ]]; then
+  ALL_COMMITS_MUST_PASS=1
+else
+  ALL_COMMITS_MUST_PASS=0
+fi
+
+# We need all the commits in the PR to be available
+. tools/ci/actions/deepen-fetch.sh
+
+# Display failing commits in red for PRs and yellow for branches (error/warning)
+if ((ALL_COMMITS_MUST_PASS)); then
+  COLOR='31'
+else
+  COLOR='33'
+fi
+
+CI_SCRIPT='tools/ci/actions/check-configure.sh'
+PATHS='configure\|configure\.ac\|VERSION\|aclocal\.m4\|build-aux/.*'
+
+# $1 - commit to checkout files from
+# $2 - range of commits to diff
+# When testing a single commit, $1 and $2 will be the same; when validating the
+# tip of a branch, $1 will be HEAD and $2 will be the range of commits in the
+# branch.
+CheckTree () {
+  RET=0
+  COMMIT="$1"
+  COMMITS_TO_SEARCH="$2"
+  if git diff-tree --diff-filter=d --no-commit-id --name-only -r \
+       "$COMMITS_TO_SEARCH" | grep -qx "$PATHS\|$CI_SCRIPT"; then
+    git checkout -qB return
+    git checkout -q "$COMMIT"
+    mv configure configure.ref
+    make -s configure
+    if diff -q configure configure.ref >/dev/null ; then
+      echo -e "$COMMIT: \e[32mconfigure.ac generates configure\e[0m"
+    else
+      RET=1
+      echo -e \
+        "$COMMIT: \e[${COLOR}mconfigure.ac doesn't generate configure\e[0m"
+    fi
+    mv configure.ref configure
+    git checkout -q return
+  fi
+  return $RET
+}
+
+RESULT=1
+for commit in $(git rev-list "$MERGE_BASE..$PR_HEAD" --reverse); do
+  if CheckTree "$commit" "$commit"; then
+    if ((!ALL_COMMITS_MUST_PASS)); then
+      RESULT=1
+    fi
+  else
+    RESULT=0
+  fi
+done
```

Comment:
It took me a while to understand this code (the corresponding loop in `check-typo` is easier to read as it does not return any value, failures are signalled from the inner function call itself). Could this be written (roughly) as follows?

```sh
RESULT=1 # 1 for success, 0 for check failure
if ((!ALL_COMMIT_MUST_PASS)); then
  for commit in $(git rev-list "$MERGE_BASE..$PR_HEAD" --reverse); do
    if ! CheckTree "$commit" "$commit"; then RESULT=0; fi
  done
else
  if ! CheckTree "$PR_HEAD" "$PR_HEAD"; then RESULT=0; fi
fi
```

(If I understand correctly, we never access $2 from CheckTree, so the code could be simplified, but maybe you had in mind to use a commit-range again for something in the future?)

Reply:
I pushed more comments (this is very useful - there's every chance I'll forgot what I did!).

The configure check always loops over all the commits (which is different from your case, and definitely different from the check-typo case, which is why I renamed the variable to `ALL_COMMITS_MUST_PASS` rather than `CHECK_ALL_COMMITS`). However, if `ALL_COMMITS_MUST_PASS=0` then on each commit which passes we can _reset_ `RESULT` to `1` which allows it to test that the _last_ commit which modified `configure` did so correctly, while also _warning_ about any broken ones in between.

Reply:
Ah, I had missed that we would still get the warning for previous commits, thanks.

(One thing that is slightly non-obvious is that this code works because we pass `--reverse` to get the commits from oldest to newest, so the "whoever writes to the flag last wins" semantics correctly implement the meaning of the negation of ALL_COMMIT_MUST_PASS, which is actually IT_SUFFICES_THAT_THE_LAST_COMMIT_PASSES.) 

(More concisely:
```
# We traverse the commit in commit order; if ALL_COMMITS_MUST_PASS is not set,
# the success of the head of the branch, traversed last, will override any previous failure.
```
)

Reply:
Done (with a slight clarification that it's the most recent commit, rather than necessarily HEAD)